Sources_v3/StreamChat/ChatClient.swift

```diff
@@ -271,54 +264,214 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
+        tokenProvider: _TokenProvider<ExtraData>,
         workerBuilders: [WorkerBuilder],
         eventWorkerBuilders: [EventWorkerBuilder],
         environment: Environment
     ) {
+        log.assert(Thread.isMainThread, "ChatClient.init must be called on main thread.")
+
         self.config = config
+        self.tokenProvider = tokenProvider
         self.environment = environment
         self.workerBuilders = workerBuilders
         self.eventWorkerBuilders = eventWorkerBuilders
-        
-        createBackgroundWorkers()
+        currentUserId = databaseContainer.viewContext.currentUser()?.user.id
+
         createWebSocketClient()
+        reloadUserIfNeeded()
     }
     
     deinit {
-        _connectionIdWaiters.mutate { waiters in
-            waiters.forEach { $0(nil) }
-            waiters.removeAll()
-        }
+        completeConnectionIdWaiters(passing: nil)
+        completeTokenWaiters(passing: nil)
     }
-    
-    // TODO: Not used & tested yet -> CIS-224
-    func refreshToken(completion: @escaping (Error?) -> Void) {
-        log.assert(config.tokenProvider != nil, "You can't call `refreshToken` when `Config.tokenProvider` is nil.")
-        
-        config.tokenProvider?(config.apiKey, currentUserId, { [weak self] (token) in
-            guard let token = token else {
-                completion(ClientError.MissingToken("Can't connect because `TokenProvider` didn't return a valid token."))
-                return
-            }
-            self?.currentToken = token
-            completion(nil)
-        })
-    }
-    
+}
+
+private extension _ChatClient {
     func createBackgroundWorkers() {
         backgroundWorkers = workerBuilders.map { builder in
             builder(self.databaseContainer, self.apiClient)
         } + eventWorkerBuilders.map { builder in
             builder(self.databaseContainer, self.eventNotificationCenter, self.apiClient)
         }
     }
-    
+
+    func completeConnectionIdWaiters(passing connectionId: String?) {
+        _connectionIdWaiters.mutate { waiters in
+            waiters.forEach { $0(connectionId) }
+            waiters.removeAll()
+        }
+    }
+
+    func completeTokenWaiters(passing token: Token?) {
+        _tokenWaiters.mutate { waiters in
+            waiters.forEach { $0(token) }
+            waiters.removeAll()
+        }
+    }
+
     /// Before CIS-555, `webSocketClient` was initialized in `createBackgroundWorkers` in `init` since it was passed
     /// to the background workers. Now, background workers do not access `webSocketClient` and it's not being
     /// initialized in `init`, so we initialize it here explicitly.
-    private func createWebSocketClient() {
+    func createWebSocketClient() {
         _ = webSocketClient
     }
+
+    func prepareEnvironment(newToken: Token, completion: @escaping (Error?) -> Void) {
+        // Update web-socket endpoint.
+        webSocketClient?.connectEndpoint = .webSocketConnect(userId: newToken.userId)
```

Comment:
This doesn't need to be done if the new user id is same as the current one.

Reply:
This does need to be done (at least in current PR shape) since we removed the 
```swift
if let currentUserId = currentUserId {
 webSocketClient?.connectEndpoint = .webSocketConnect(userId: currentUserId)
}
```
from `lazy var webScoketClient` so initially `webScoketClient.connectEndpoint == nil` and has to be set somewhere.

_In `webSocketClient` we now have `createOrReuseEngine` which builds the new engine only if the request is different (e.g. the `user_id` is changed) so we can set the endpoint multiple times without extra job._

Reply:
To me setting the endpoint when web-socket is created and resetting if the user has changed makes more sense.